src/celeritas/optical/model/AbsorptionModel.cc

```diff
@@ -0,0 +1,62 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/optical/model/AbsorptionModel.cc
+//---------------------------------------------------------------------------//
+#include "AbsorptionModel.hh"
+
+#include "corecel/Assert.hh"
+#include "celeritas/io/ImportOpticalMaterial.hh"
+#include "celeritas/optical/detail/MfpBuilder.hh"
+
+namespace celeritas
+{
+namespace optical
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct the model from imported data.
+ */
+AbsorptionModel::AbsorptionModel(ActionId id, ImportedModelAdapter imported)
+    : Model(id, "absorption", "interact by optical absorption")
+    , imported_(std::move(imported))
+{
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Build the mean free paths for the model.
+ */
+void AbsorptionModel::build_mfps(detail::MfpBuilder& build) const
+{
+    for (auto mat : range(OpticalMaterialId{imported_.num_materials()}))
+    {
+        build(imported_.mfp(mat));
+    }
+}
```

Comment:
I think `build_mfp(OpticalMaterialId, MfpBuilder&)` is nicer because it makes `MfpBuilder` less stateful and less coupled to optical models. Looping in the `build_mfps` was a byproduct of not having an immediate way to get the number of optical materials in `optical::PhysicsParams::build_mfps`, but that should be easily resolved.